commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java

```diff
@@ -0,0 +1,475 @@
+     * generating the convex hull itself.
+     * <p>
+     * The complexity of the operation is O(n), and may greatly improve the time it
+     * takes to construct the convex hull afterwards, depending on the point
+     * distribution.
+     *
+     * @see <a href=
+     *      "http://en.wikipedia.org/wiki/Convex_hull_algorithms#Akl-Toussaint_heuristic">
+     *      Akl-Toussaint heuristic (Wikipedia)</a>
+     */
+    public static final class Builder {
+
+        /** Corner of quadrilateral with minimal x coordinate. */
+        private Vector2D minX;
+
+        /** Corner of quadrilateral with maximal x coordinate. */
+        private Vector2D maxX;
+
+        /** Corner of quadrilateral with minimal y coordinate. */
+        private Vector2D minY;
+
+        /** Corner of quadrilateral with maximal y coordinate. */
+        private Vector2D maxY;
+
+        /** Collection of all remaining candidates for a convex hull. */
+        private final Collection<Vector2D> candidates;
+
+        /** A precision context for comparing points. */
+        private final Precision.DoubleEquivalence precision;
+
+        /** Indicates if collinear points on the hull shall be present in the output.
+         * If {@code false}, only the extreme points are added to the hull.
+         */
+        private final boolean includeCollinearPoints;
+
+        /** Return a {@link Builder} instance configured with the given precision
+         * context. The precision context is used when comparing points.
+         *
+         * @param builderPrecision       precision context to use when building a convex
+         *                               hull from raw vertices; may be null if raw
+         *                               vertices are not used.
+         * @param includeCollinearPoints whether collinear points shall be added as hull
+         *                               vertices
+         */
+        public Builder(final boolean includeCollinearPoints, final Precision.DoubleEquivalence builderPrecision) {
+            this.precision = builderPrecision;
+            this.includeCollinearPoints = includeCollinearPoints;
+            candidates = EuclideanCollections.pointSet2D(builderPrecision);
+        }
+
+        /** Appends the given point to a collection of possible hull points, if and only
+         * if the given point is outside of a constructed quadrilateral of extreme properties.
+         *
+         * @param point a given point.
+         * @return this instance.
+         */
+        public Builder append(Vector2D point) {
+
+            // Checks if the given point supersedes one of the corners.
+            checkCorners(point);
+
+            // Only proceed if the quadrilateral is complete.
+            if (candidates.size() < 4) {
+                return this;
+            }
+
+            final List<Vector2D> quadrilateral = buildQuadrilateral(minY, maxX, maxY, minX);
```

Comment:
It may be useful to cache the quadrilateral. Modification of the `checkCorners` method can be done to remove the cached quadrilateral if the point supersedes one of the corners.

Reply:
Or maybe get rid of the list at all. The way I see it. The purpose of the list is to count how many of the points are truly distinct and to go around all sides in an ordered manner. Since we only have 4 corners at all and only proceed if there are  at least 3 sides it would be possible to get rid of the list and just to count and then replace the loop with a method that checks either all three or four sides.

Reply:
It isn't really a new approach, but maybe more the fact, that the loop has only 3-4 iterations which could be straight forward to code. Since we only check the sides [minY, maxX], [maxX, maxY], [maxY, minX], [minX, minY]  for any given point, whereas we skip any side if the two points align, which can happen only once. I guess one could implement this logic entirely without a loop. 
I guess the simplest way is still to just cache the list.